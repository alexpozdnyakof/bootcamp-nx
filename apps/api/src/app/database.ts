import { Database } from 'sqlite3'

const dbFileName =
	process.env.NODE_ENV === 'test' ? `database-test` : `database`

const database = new Database(`${__dirname}/${dbFileName}.sqlite`, err => {
	if (err) throw new Error(err.message)
})

function migrate() {
	console.log({ __dirname })

	database.exec(
		'CREATE TABLE IF NOT EXISTS projects (id INTEGER PRIMARY KEY AUTOINCREMENT,title VARCHAR(200) NOT NULL,description TEXT NULL)'
	)

	if (process.env.NODE_ENV === 'test') {
		console.log('test')
		addMigrationData()
	}
}

function addMigrationData() {
	console.log({ __dirname })
	database.exec(
		`INSERT OR REPLACE INTO projects VALUES (1, 'ホームページのリニューアル', '新しいフレッシュホームページの制作過程はこちら'), (2, 'チェックアウトフォームのリファクタリング', '新しいチェックアウト フォームでユーザー エクスペリエンスを向上'), (3, '実験プロジェクト', null)`
	)
}

export { database, migrate, addMigrationData }

// 	'CREATE TABLE IF NOT EXISTS projects (id INTEGER PRIMARY KEY AUTOINCREMENT,title VARCHAR(200) NOT NULL,description TEXT NULL)'
// `INSERT OR REPLACE INTO projects VALUES (1, 'ホームページのリニューアル', '新しいフレッシュホームページの制作過程はこちら'), (2, 'チェックアウトフォームのリファクタリング', '新しいチェックアウト フォームでユーザー エクスペリエンスを向上'), (3, '実験プロジェクト', null)`
