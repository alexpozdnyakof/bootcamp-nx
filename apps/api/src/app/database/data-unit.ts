type UniqueId = number

type UpdatedCreatedTime = {
	created: number
	updated: number
}

interface EntityPropsMap {
	task: {
		text: string
		done: boolean
	}
	task_list: {
		title: string
	}
	project: {
		title: string
	}
}

type EntityType = 'task' | 'task_list' | 'project'

type DataUnit<T extends EntityType> = {
	id: UniqueId
	parentId: UniqueId
	type: T
	properties: EntityPropsMap[T] & UpdatedCreatedTime
	children: Array<UniqueId>
}

type EntityUnit = DataUnit<'task'> | DataUnit<'task_list'> | DataUnit<'project'>

type AbstractStore = Array<EntityUnit>

type DataUnitDTO = Pick<
	DataUnit<keyof EntityPropsMap>,
	'parentId' | 'children' | 'type'
> & {
	properties: Omit<
		DataUnit<keyof EntityPropsMap>['properties'],
		'created' | 'updated'
	>
}

type DataUnitUpdateValue = Partial<
	Pick<DataUnitDTO, 'properties' | 'children' | 'parentId'>
> &
	Pick<EntityUnit, 'id'>

export {
	AbstractStore,
	EntityType,
	UniqueId,
	EntityUnit,
	DataUnitDTO,
	DataUnit,
	DataUnitUpdateValue,
	UpdatedCreatedTime,
}
