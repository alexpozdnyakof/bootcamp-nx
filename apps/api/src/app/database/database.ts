import { Database } from 'sqlite3'

import * as fs from 'fs'
import * as path from 'path'

const dbFileName = process.env.NODE_ENV === 'test' ? `test` : `database`

const database = new Database(`${__dirname}/${dbFileName}.sqlite`, err => {
	if (err) throw new Error(err.message)
})
const sqlUrl =
	process.env.NODE_ENV === 'test'
		? path.join(__dirname, '../../sql')
		: path.join(__dirname, '/sql')

function migrate() {
	console.log({ __dirname })

	database.exec(fs.readFileSync(sqlUrl.concat('/projects.sql')).toString())

	if (process.env.NODE_ENV === 'test') {
		console.log('test')
		seeds()
	}
}

function seeds() {
	console.log({ __dirname })
	database.exec(
		fs.readFileSync(sqlUrl.concat('/projects-migration.sql')).toString()
	)
}

export { database, migrate }

// 'CREATE TABLE IF NOT EXISTS projects (id INTEGER PRIMARY KEY AUTOINCREMENT,title VARCHAR(200) NOT NULL,description TEXT NULL)'
// `INSERT OR REPLACE INTO projects VALUES
// (1, 'ホームページのリニューアル',
// '新しいフレッシュホームページの制作過程はこちら'),
// (2, 'チェックアウトフォームのリファクタリング', '新しいチェックアウト フォームでユーザー エクスペリエンスを向上'), (3, '実験プロジェクト', null)`
