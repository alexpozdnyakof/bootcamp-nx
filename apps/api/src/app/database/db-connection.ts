import knex from 'knex'
import knexConfig from '../../knexfile'
const dbConnection = () => {
	const envConfig = knexConfig[process.env.NODE_ENV ?? 'development']
	return knex(envConfig)
}

export default dbConnection()
