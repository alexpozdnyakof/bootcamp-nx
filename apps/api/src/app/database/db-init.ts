import { Database } from 'sqlite3'
import { readFileSync, readdirSync } from 'node:fs'
import { join } from 'path'

function readSqlDirectory(subdir: string): Array<string> {
	const path = join(
		__dirname,
		process.env.NODE_ENV === 'test' ? '../../sql' : '/sql',
		subdir
	)
	return readdirSync(path).map(fileName =>
		readFileSync(join(path, '/', fileName)).toString()
	)
}

function migrate(dbInstance: Database) {
	const migrations = readSqlDirectory('/migrations')
	migrations.map(sqlString => dbInstance.exec(sqlString))
}

function seed(dbInstance: Database) {
	const seeds = readSqlDirectory('/seeds')
	seeds.map(sqlString => dbInstance.exec(sqlString))
}

export function InitSqlite() {
	const isTest = process.env.NODE_ENV === 'test'
	const dbpath = isTest ? ':memory:' : join(__dirname, '/database.sqlite')

	const database = new Database(dbpath, err => {
		if (err) throw new Error(err.message)
	})

	migrate(database)
	if (process.env.NODE_ENV === 'test') {
		seed(database)
	}

	return database
}

export default InitSqlite()
