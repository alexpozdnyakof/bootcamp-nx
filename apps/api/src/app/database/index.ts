export { default as DataModel } from './data-model'
export * from './data-unit'
export { default as Database } from './db'
