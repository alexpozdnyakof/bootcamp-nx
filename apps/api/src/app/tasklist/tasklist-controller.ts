import { ApiTask } from '@bootcamp-nx/api-interfaces'
import { Router, Response } from 'express'
import { CreateTask } from '../task/task'
import { TaskListValue } from './tasklist'
import { TaskListRepo } from './tasklist-repo'

const TasklistRouter = Router()
const TaskListModel = TaskListRepo()
const TasklistRouterPrefix = 'tasklist'

TasklistRouter.get('/:id', async (request, response) => {
	const id = Number(request.params.id)
	try {
		const tasklist = await TaskListModel.GetOne(id)
		response.status(200).send(tasklist)
	} catch (error) {
		response.status(404).send({ message: error.message })
	}
})

TasklistRouter.get(
	'/:id/tasks',
	async (request, response: Response<ApiTask[] | { message: string }>) => {
		const id = Number(request.params.id)
		try {
			const tasks = (await TaskListModel.GetRelatedTasks(id)).map(task =>
				CreateTask(task)
			)
			response.status(200).send(tasks)
		} catch (error) {
			response.status(404).send({ message: error.message })
		}
	}
)

TasklistRouter.post('/', async (req, res) => {
	try {
		const dto = TaskListValue.check(req.body)
		const result = await TaskListModel.Add(dto)
		res.status(201).send(result)
	} catch (error) {
		res.status(400).send({ message: error.message })
	}
})

TasklistRouter.delete('/:id', async (req, res) => {
	const id = Number(req.params.id)
	try {
		const tasklist = await TaskListModel.GetOne(id)
		if (typeof tasklist == 'undefined') throw new Error('Not Found')
		await TaskListModel.Delete(id)
		res.status(204).send()
	} catch (error) {
		res.status(400).send({ message: error.message })
	}
})

TasklistRouter.put('/:id', async (req, res) => {
	const id = Number(req.params.id)
	try {
		const dto = TaskListValue.check(req.body)
		await TaskListModel.Update(id, dto)
		res.status(204).send()
	} catch (error) {
		res.status(400).send({ message: error.message })
	}
})

export { TasklistRouter as TasklistController, TasklistRouterPrefix }
