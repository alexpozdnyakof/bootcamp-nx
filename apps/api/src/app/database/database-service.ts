import database from './database'

type DbResult<T> = {
	status: 'success' | 'failed'
	result: T
}

export function DatabaseService() {
	return Object.freeze({
		async All<T>(
			query: string,
			params?: Array<string | number | boolean>
		): Promise<T> {
			return new Promise((resolve, reject) => {
				database.all(query, params ?? [], (err, result: T) => {
					if (err) reject(err)
					resolve(result)
				})
			})
		},
		async Get<T>(
			query: string,
			params?: Array<string | number | boolean>
		): Promise<T> {
			return new Promise((resolve, reject) => {
				database.get(query, params ?? [], (err, result: T) => {
					if (err) reject(err)
					resolve(result)
				})
			})
		},
		async Run<T>(
			query: string,
			params?: Array<string | number | boolean>
		): Promise<T> {
			return new Promise((resolve, reject) => {
				database.run(query, params ?? [], (err, result: T) => {
					if (err) reject(err)
					resolve(result)
				})
			})
		},
	})
}

export default DatabaseService()
