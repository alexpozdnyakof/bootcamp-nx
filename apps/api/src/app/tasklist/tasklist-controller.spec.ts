import { json } from 'body-parser'
import express from 'express'
import request from 'supertest'
import { database } from '../database'
import { TasklistController } from './tasklist-controller'
describe('TaskListController', () => {
	const App = express()
	App.use(json())
	App.use('/', TasklistController)

	beforeAll(async () => {
		await database.migrate.up({
			name: '20221129145937_create_tasklist_table.ts',
		})
		await database.migrate.up({
			name: '20221129145851_create_task_table.ts',
		})
		await database.migrate.up({
			name: '20221202114516_create_task_to_tasklist.ts',
		})
		await database.migrate.up({
			name: '20221130102828_create_tasklist_to_project.ts',
		})
		await database.seed.run({ specific: '02-tasklist.ts' })
		await database.seed.run({ specific: '03-task.ts' })
	})

	it('should return all related tasks', async () => {
		const response = await request(App).get('/1/tasks')
		expect(response.status).toBe(200)

		expect(response.body).toMatchSnapshot()
	})
	it('should create new one tasklist', async () => {
		const response = await request(App)
			.post('/')
			.set('Accept', 'application/json')
			.send({ title: '新しい計画', description: '簡単な説明' })

		expect(response.status).toBe(201)
	})

	it('should delete tasklist', async () => {
		const response = await request(App).delete('/3')
		expect(response.status).toBe(204)
	})

	it('should update tasklist', async () => {
		const response = await request(App).put('/1').send({
			title: '新しい計画',
			description: '簡単な説明',
		})

		expect(response.status).toBe(204)
	})
})
