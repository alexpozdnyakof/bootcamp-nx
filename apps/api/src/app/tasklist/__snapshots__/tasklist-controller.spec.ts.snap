// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskListController should return all related tasks 1`] = `
Array [
  Object {
    "created": "2022-11-29 15:31:37",
    "done": false,
    "id": 1,
    "title": "血液レポートのグラフが空白になっている",
    "type": "task",
    "updated": "2022-11-29 15:31:37",
  },
  Object {
    "created": "2022-11-29 15:31:37",
    "done": true,
    "id": 2,
    "title": "無効にする|| ユーザーがアカウントを無効にできない",
    "type": "task",
    "updated": "2022-11-29 15:31:37",
  },
]
`;
